vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/LinkManager.java

```diff
@@ -302,7 +304,79 @@ public IModuleLinkDisplay createModuleLink(String title, ISpace source, float po
+        link.setName(title);
+        IModule module = new Module();
+        module.setId(linkedModuleId);
+        link.setModule(module);
+        moduleLinkRepo.save((ModuleLink) link);
+        Optional<ModuleLinkDisplay> moduleLinkOptional = moduleLinkDisplayRepo.findById(moduleLinkDisplayId);
+        IModuleLinkDisplay display = moduleLinkOptional.get();
+        display.setPositionX(positionX);
+        display.setPositionY(positionY);
+        display.setRotation(rotation);
+        moduleLinkDisplayRepo.save((ModuleLinkDisplay) display);
+        return display;
+    }
+    
+    
+    @Override
+    public ISpaceLinkDisplay editSpaceLink(String title, ISpace source, float positionX, float positionY,
+            int rotation, String linkedSpaceId, String spaceLinkLabel, byte[] linkImage, String spaceLinkIdValueEdit, String spaceLinkDisplayId,
+            String imageFilename) throws ImageCouldNotBeStoredException, SpaceDoesNotExistException {
+        
+        source = spaceManager.getSpace(source.getId());
+        if (source == null) {
+            throw new SpaceDoesNotExistException();
+        }
+        Optional<SpaceLink> linkOptional = spaceLinkRepo.findById(spaceLinkIdValueEdit);
+        ISpaceLink link = linkOptional.get();
+        link.setName(title);
+        ISpace space = new Space();
+        space.setId(linkedSpaceId);
+        link.setTargetSpace(space);
+        spaceLinkRepo.save((SpaceLink) link);
+        Optional<SpaceLinkDisplay> spaceLinkOptional = spaceLinkDisplayRepo.findById(spaceLinkDisplayId);
+        ISpaceLinkDisplay display = spaceLinkOptional.get();
+        display.setPositionX(positionX);
+        display.setPositionY(positionY);
+        display.setRotation(rotation);
+        spaceLinkDisplayRepo.save((SpaceLinkDisplay) display);
+        return display;
+    }
+    
+    @Override
+    public IExternalLinkDisplay editExternalLink(String title, ISpace source, float positionX, float positionY,
+            String externalLink, byte[] linkImage, String imageFilename, String externalLinkIdValueEdit, String externalLinkDisplayId ) throws ImageCouldNotBeStoredException, SpaceDoesNotExistException{
+        source = spaceManager.getSpace(source.getId());
+        if (source == null) {
+            throw new SpaceDoesNotExistException();
+        }
+        Optional<ExternalLink> linkOptional = externalLinkRepo.findById(externalLinkIdValueEdit);
+        IExternalLink link = linkOptional.get();
+        link.setName(title);
+        link.setExternalLink(externalLink);
+        externalLinkRepo.save((ExternalLink) link);
+        Optional<ExternalLinkDisplay> externalLinkOptional = externalLinkDisplayRepo.findById(externalLinkDisplayId);
+        IExternalLinkDisplay display = externalLinkOptional.get();
+        display.setPositionX(positionX);
+        display.setPositionY(positionY);
+        externalLinkDisplayRepo.save((ExternalLinkDisplay) display);
+        return display;
```

Comment:
looks like this the same code three times. there should be one method that does all the common stuff and the individual methods call this one method (this will be easier to do with the above described changes).

Reply:
display is an instance of different objects in each of the functions.

Reply:
but they inherit from the same superclass, don't they?

Reply:
Yes, converted into single function,